### PR TITLE
[Feature] Implement min_by aggregate function

### LINKED ISSUE #3288: Is there a function like argMax() argMin() in Clickhouse or min_by() max_by() in presto？
比如在分组统计中，我要获取时间戳最近的一条数据，或者获取时间戳最早的一条数据 ，ck中可以使用argMax(xx,timestamp),argMin(xx,timestamp)来解决
presto中的max_by()，min_by()也可以解决，不知是否有相关类似函数？或者有计划在哪个版本提供类似函数支持？

### PR DESCRIPTION
**Pull request by ucasfl**
## What type of PR is this：
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #3288

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [x] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 22 / 25 (88.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/FunctionAnalyzer.java|8|11|72.73%|[230, 249, 260]|
|:large_blue_circle:|com/starrocks/catalog/FunctionSet.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/DecimalV3FunctionAnalyzer.java|6|6|100.00%|[]|

**Comment by ucasfl** [comment]
@satanson Hi, could you help with review this?

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/FunctionAnalyzer.java [hunk 1]
```diff
@@ -245,6 +244,25 @@ private static void analyzeBuiltinAggFunction(FunctionCallExpr functionCallExpr)
...
+            }
+
+            fnParams.setIsDistinct(false);  // DISTINCT is meaningless here
+
+            Type sortKeyType = functionCallExpr.getChild(1).getType();
+            if (!sortKeyType.canApplyToNumeric()) {
+                throw new SemanticException(Type.ONLY_METRIC_TYPE_ERROR_MSG);
+            }
+
+            return;
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/analyzer/FunctionAnalyzer.java:263]: merge repeat code?

**Comment by Seaven** [comment]
and you need `git commit -s` to pass the DCO check, you can try `git commit --amend -s ` & `git push -f` again

**Comment by ucasfl** [comment]
> and you need `git commit -s` to pass the DCO check, you can try `git commit --amend -s ` & `git push -f` again

done.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 14 / 14 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/FunctionSet.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/DecimalV3FunctionAnalyzer.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/FunctionAnalyzer.java|1|1|100.00%|[]|

**Comment by ucasfl** [comment]
I have commit with sign but DCO check also failed.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 12 / 14 (85.71%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/FunctionAnalyzer.java|3|5|60.00%|[240, 241]|
|:large_blue_circle:|com/starrocks/catalog/FunctionSet.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/DecimalV3FunctionAnalyzer.java|1|1|100.00%|[]|

**Review by Seaven — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/catalog/FunctionSet.java [hunk 2]
```diff
@@ -782,8 +783,17 @@ private void initAggregateBuiltins() {
...
-                        Lists.newArrayList(t1, t), t1, Type.VARBINARY, true, true, false));
+                addBuiltin(AggregateFunction.createBuiltin(
+                        MAX_BY, Lists.newArrayList(t1, t), t1, Type.VARBINARY, true, true, false));
+            }
+
+            // MIN_BY
+            for (Type t1 : Type.getSupportedTypes()) {
+                if (t1.isFunctionType() || t1.isNull() || t1.isChar() || t1.isPseudoType()) {
+                    continue;
+                }
```
Comment (fzhedu) [inline fe/fe-core/src/main/java/com/starrocks/catalog/FunctionSet.java:794]: can it support nested type? like array, map, struct?

**Comment by satanson** [comment]
@ucasfl min_by and max_by are dual forms, so suggest to use the same suite of  aggregation implementation,  and add a template bool argment to select min_by and max_by specialization.

**Comment by ucasfl** [comment]
> @ucasfl min_by and max_by are dual forms, so suggest to use the same suite of aggregation implementation, and add a template bool argment to select min_by and max_by specialization.

@satanson I don't understand which places should be rewrite with template bool argument, `MaxByElement` and `MinByElement`?

File: be/test/exprs/agg/aggregate_test.cpp [hunk 3]
```diff
@@ -707,6 +707,62 @@ TEST_F(AggregateTest, test_maxby) {
...
+    Columns columns;
+    columns.emplace_back(int_column);
+    columns.emplace_back(varchar_column);
+    std::vector<const Column*> raw_columns;
+    raw_columns.resize(columns.size());
+    for (int i = 0; i < columns.size(); ++i) {
+        raw_columns[i] = columns[i].get();
+    }
+    func->update_batch_single_state(ctx, int_column->size(), raw_columns.data(), aggr_state->state());
+    func->finalize_to_column(ctx, aggr_state->state(), result_column.get());
```
Comment (fzhedu) [inline be/test/exprs/agg/aggregate_test.cpp:730]: how about testing `merge`, `serialize_to_column` and `convert_to_serialize_format` in aggregation functions?

**Review by fzhedu — COMMENTED** [review]
add doc pls

File: fe/fe-core/src/main/java/com/starrocks/catalog/FunctionSet.java (hunk 2 above)
Comment (ucasfl) [reply fe/fe-core/src/main/java/com/starrocks/catalog/FunctionSet.java:794]: Looks like it does not support, and there is no <> operator for array, map, struct.

Comment (fzhedu) [reply fe/fe-core/src/main/java/com/starrocks/catalog/FunctionSet.java:794]: add `|| t1.isComplexType()`?

**Review by Seaven — DISMISSED** [review]

**Review by evelynzhaojie — APPROVED** [review]

**Comment by Seaven** [comment]
run all

**Review by satanson — APPROVED** [review]

**Pull request merged**